std/experimental/perpetual.d

```diff
@@ -0,0 +1,653 @@
+		static if(dynamic)
+		{
+			_heap=shMem(path, mode);
+			enforce(_heap.length >= Element.sizeof, _tag~": file is too small");
+			_value=cast(Element[]) _heap[0.._heap.length];
+			// initialization
+			static if(isAssignable!Element)
+			{
+				if(master)
+					foreach(i, ref x; _value[initIndex..$])
+						emplace(&x, arg);
+			}
+			else
+			{
+				static assert(arg.length == 0
+					, _tag~": attempt to initialize immutable memory");
+			}
+		}
+		// scalar value
+		else
+		{
+			_heap=shMem(path, Element.sizeof, mode);
+			enforce(_heap.length >= Element.sizeof, _tag~": file is too small");
+			_value=cast(Element*) _heap[].ptr;
+
+			// initialization
+			static if(isAssignable!Element)
+			{
+				if(master)
+					emplace(_value, arg);
+			}
+			else
+			{
+				static assert(arg.length == 0
+					, _tag~": attempt to initialize immutable object");
+			}
+		}
+	}
+
+/**
+ * Open file and map dynamic array to it.
+ * Creates array of requested length, file is extended if necessary
+ */
+	this(Arg...)(size_t len, string path, Arg arg)
+	if(dynamic)
+	{
+		immutable size_t size=len*Element.sizeof;
+		
+		_heap=shMem(path, size, mode);
+		_value=cast(Element[]) _heap[0..size];
+		// initialization
+		static if(isAssignable!Element)
+		{
+			if(master)
+				foreach(ref x; _value[initIndex..$])
+					emplace(&x, arg);
+		}
+		else
+		{
+			static assert(arg.length == 0
+				, _tag~": attempt to initialize immutable array");
+		}
+	}
+ 	
+
+ // get reference to wrapped object.
+	alias Ref this;
+
+}
+
+
+///
+unittest {
+	import std.stdio;
+	import std.conv;
+	import std.string;
+	import std.file : remove, exists;
+	import core.sys.posix.sys.stat;
+	import std.file : deleteme;
+
+	// custom data examples
+	struct A { int x; };
+	class B {};
+	enum Color { black, red, green, blue, white };
+
+	string[] file;
+	foreach(i; 0..8) file~=deleteme~to!string(i);
+	scope(exit) foreach(f; file[]) if(exists(f)) remove(f);
+
+	/// Part 1: create mapped variables
+	{
+		// simle int variable initialized with default value
+		auto p0=Perpetual!int(file[0]);
+		assert(p0 == int.init);
+		p0=7;
+		assert(p0 == 7);
+
+		// single double value initialized in ctor
+		// , would throw if the file did exist.
+		auto p1=Perpetual!double(file[1], 2.71828);
+		assert(p1 == 2.71828);
+		p1=3.14159;
+
+		// struct, initialized in ctor
+		auto p2=Perpetual!A(file[2], 22);
+		assert(p2.x == 22);
+		
+		// static array of integers, assignable
+		auto p3=Perpetual!(int[5])(file[3]);
+		assert(p3[0] == int.init);
+		p3=[1,3,5,7,9];
+		assert(p3[0] == 1);
+
+		// enum, initialized in ctor
+		auto p4=Perpetual!Color(file[4], Color.red);
+		assert(p4 == Color.red);
+		
+
+		// character string
+		auto p5=Perpetual!(char[32])(file[5]);
+		p5="hello world";
+
+		// second order static array
+		auto p8=Perpetual!(char[3][5])(file[6]);
+		p8[]="..."; p8[1]="one"; p8[2]="two";
+
+
+		/// Compile time errors
+		// Perpetual!(char*)("?");        //ERROR: "char* is reference type"
```

Comment:
I believe you can test this via `static assert(!is(typeof(Perpetual!(char*)("?"))));` This applies the others as well.